### DISCUSSION TITLE
About the computational graph mechanism of flax

### DISCUSSION THREAD

**Discussion by cjLiu666** [google/flax#1682] [Q&A] [answered]
As we know, pytorch is based on dynamic graph and tensorflow is based on static graph.
So what about flax? I run the CNN example, and insert print statement in the __call__ method, the result is the same as pytorch,
the media result can be seen. Is it means flax is based on dynamic graph, like pytorch? 
Thanks for your reply.

**Comment by marcvanzee** [comment, marked as answer]
If you want to compile a static computation graph for a program in Flax, you generally use the [jax.jit](https://jax.readthedocs.io/en/latest/jax.html?highlight=jit#jax.jit) transformation provided by JAX, which will JIT compile a JAX-compatible function. If you look at some of the [Flax examples](https://github.com/google/flax/tree/main/examples), you see most of our `train_step` functions are decorated with `@jax.jit` (or `jax.pmap`, which also JIT compiles the functions but maps the first axes of the inputs to all available devices). 

In order to understand this better, I recommend reading the JAX tutorial [Just In Time Compilation with JAX](https://jax.readthedocs.io/en/latest/jax-101/02-jitting.html). Actually I highly recommend going through the all [JAX 101 tutorials](https://jax.readthedocs.io/en/latest/jax-101/index.html). Having a good understanding of JAX is really beneficial if you want to write code in Flax.

**Reply by cjLiu666** [reply]
Thanks for your reply. I have learned some JAX tutorials, and known some about jit. My understanding is flax is dynamic computation graph, if not be decorated with @jax.jit，otherewise it becomes a static computation graph. Is my understanding right? Thanks again for your reply.

**Reply by marcvanzee** [reply]
Yes that seems to be more or less correct. However note that we rely heavily on JIT compilation, and if you don't use this the code may run very slowly and you may run out of memory.

**Reply by cjLiu666** [reply]
I get it, thanks a lot. I can understand the speed may be very slow, But why the memory is run out? Does JIT use a memory optimization technology? And flax do not achieve this?

**Reply by marcvanzee** [reply]
(Updated my reply after remark from @cjLiu666 below)

JIT will automatically compile the program for the backend that is available, and returns an optimized version of the input function optimized for just-in-time compilation. So this means it can optimise the graph by, for instance, fusing certain operations, or doing other optimizations that reduce the memory usage.

**Reply by cjLiu666** [reply]
Thanks a lot! I am a fresh man for deep learning and JAX. You said that If don't JIT compile, the code will run on CPU. So I run a small example in the Colab，but see the different result. Set the Colab with GPU and run the code without JIT，the data and computation is moved to GPU automatically（I don't set the data device in code Explicitly). So I think JIT just generates a optimizal media represent, such as operation fusion, and compile the program for the available backend. But without it, we can also run our code on GPU, Am I right? Or I misuderstand anything? 
![屏幕截图 2021-11-25 110711](https://user-images.githubusercontent.com/29277731/143374331-bf3206ab-b338-4c51-871e-7f74ec05f6ae.png)
![屏幕截图 2021-11-25 110740](https://user-images.githubusercontent.com/29277731/143374335-ccef74c3-9d64-43a4-87c8-b98982e7a3c0.png)

**Reply by marcvanzee** [reply]
Yes, you are totally right. I've updated my previous reply, thanks for noting!
